[ZC-1342]: Disable variable syntax in derivations

slobodan-ilic opened the pull request:
All derivation expressions should use `{"var": <alias>}` syntax. The `{"variable": <varid>}` syntax is deprecated and will not be supported in the fututre.

slobodan-ilic committed:
remove "variable" syntax for expressions

jjdelc reviewed (commented):
This is amazing work!! Thanks @slobodan-ilic

slobodan-ilic committed:
Preserve scrunch API while emitting var derivations

slobodan-ilic committed:
Address review comments for var expressions

slobodan-ilic commented on scrunch/tests/test_datasets.py:
```diff
@@ -707,7 +707,7 @@ def test_not_and(self):
                                 "function": "in",
                                 "args": [
                                     {
-                                        "variable": var1.url
```
Yep, removed the unused var1/var2 setup too in 50c268c.

slobodan-ilic commented on scrunch/datasets.py:
```diff
@@ -1162,13 +1162,13 @@ def create_fill_values(self, variables, name, alias, description=''):
             # We are in the case of a default fill, replace the -1 with the new
             # variable
-            fill_map["-1"] = {"variable": vars_by_alias[else_case["variable"]]["id"]}
+            fill_map["-1"] = {"var": else_case["var"]}
```
I kept vars_by_alias here. We no longer need it to turn fill_map entries into ids, but create_fill_values still uses it to validate source variables and keep the existing scrunch public API where callers pass variable aliases.

slobodan-ilic replied:
Clarifying after the cleanup in 1717fcb: vars_by_alias is no longer used in the fill_map construction. The fill map now uses the scrunch input alias and emits var. The vars_by_alias fetch remains earlier only for the existing categorical type validation.

slobodan-ilic commented on scrunch/expressions.py:
```diff
@@ -648,31 +636,12 @@ def _process(obj, variables):
 
-        # inspect function, then inspect variable, if multiple_response,
-        # then change in --> any
-        if 'function' in obj and 'args' in obj:
```
Good catch. Restored this path in 50c268c: MR variables parsed with in are normalized through the any handling again, with unit coverage for the var/axes payload.

slobodan-ilic commented on scrunch/tests/test_datasets.py:
```diff
@@ -595,7 +595,7 @@ def test_apply_exclusion(self):
             'expression': {
                 'function': '!=',
                 'args': [
-                    {'variable': var.url},  # Crunch needs variable URLs!
```
Removed those unused var assignments in the exclusion-filter tests in 50c268c.

slobodan-ilic commented on scrunch/expressions.py:
```diff
@@ -942,6 +858,11 @@ def _process(fragment, parent=None):
+            
+            if 'var' in fragment:
+                if 'axes' in fragment:
+                    return "{}[{}]".format(fragment['var'], fragment['axes'][0])
```
Added tests for this. prettify now covers var directly and var+axes as array syntax, and the parser/process path keeps emitting the Crunch API var/axes shape.

slobodan-ilic replied:
More concretely, the prettify coverage is:

- `TestPrettify.test_var_alias_no_dataset` for `{"var": ...}`
- `TestPrettify.test_var_square_bracket_subvariables` for `{ "var": ..., "axes": [...] }`
- `TestPrettify.test_transitional_variable_url` for the temporary old `variable` URL response shape
- `TestPrettify.test_transitional_variable_url_square_bracket_subvariables` for old subvariable URL responses

The parser/process outbound `var`/`axes` assertions are in:

- `TestExpressionProcessing.test_process_all_multiple_response`
- `TestExpressionProcessing.test_process_in_multiple_response`

slobodan-ilic commented on scrunch/expressions.py:
```diff
@@ -842,34 +791,6 @@ def prettify(expr, ds=None):
     methods = {m[1]: m[0] for m in CRUNCH_METHOD_MAP.items()}
     functions = {f[1]: f[0] for f in CRUNCH_FUNC_MAP.items()}
 
-    def _resolve_variable(var):
```
Had to bring a small resolver back temporarily. Current API responses can still return old variable URL expressions in some paths, so prettify accepts both variable and var until the server side is fully moved over.

slobodan-ilic commented on scrunch/expressions.py:
```diff
@@ -598,22 +588,21 @@ def category_ids(var_id, var_value, variables=variables):
-
             elif isinstance(var_value, str):
-                for var in variables:
+                for va, var in variables.items():
```
Agreed. I kept this scoped to the var syntax migration for now and avoided a broader category_ids refactor in this PR.

slobodan-ilic committed:
Make multiple response expansion deterministic

slobodan-ilic committed:
Keep fill values input on variable syntax

jjdelc commented on scrunch/expressions.py:
I think these docstrings need to change too

slobodan-ilic replied:
Good catch — fixed in 85963c3. Updated the example to use the `var` term to match what the parser now emits.

jjdelc reviewed (changes requested):
The function `process_expr` is not needed anymore, it exists to convert aliases to URLs, if we're now doing aliases directly. Keeping that function is unevaluated code debt that should also be removed. 

With this PR there is no reason to keep all that complex code still around generating confusion.

slobodan-ilic committed:
address JJ's comments pt 2

Clarify process_expr instead of removing it. JJ's note assumed the
function only translates aliases to URLs; the alias->URL part is in
fact already gone (earlier commits on this branch). What remains is
metadata-driven enrichment that parse_expr structurally cannot do:
subvariable -> parent+axes rewrite, category name -> id resolution,
multiple-response adaptation, any<->in swap by variable type, and
array variable expansion into or(...) chains.

Refactor for honesty:
- Module + process_expr docstrings now describe the five real tasks
  and state plainly that aliases are not converted to URLs. The
  top-of-file example payload now uses `var` to match.
- Drop dead code: base_url, the unused variable_id inner helper, the
  outer subvariables accumulator that is never written to.
- Extract the array-expansion block into _expand_array_filter with a
  docstring, and replace the loop-leaking `var` name with an
  explicitly tracked array_var.
- Section headers inside _process label each phase: walk,
  subvariable rewrite, any<->in swap, array expansion.

No behavior change — 337 unit tests still pass.

jjdelc commented on scrunch/datasets.py:
```diff
@@ -3514,4 +3514,4 @@ def execute(self, csv_file):
                 # Always delete the tmp dataset no matter what
-                tmp_ds.delete()
\ No newline at end of file
+                tmp_ds.delete()
```
Ah, closing newline, good!

jjdelc commented on scrunch/expressions.py:
```diff
@@ -33,11 +33,12 @@
+Expression objects produced by `parse_expr` are purely syntactic: they
+refer to variables by `alias` (in `var` terms) and know nothing about
+the dataset. They are not yet ready to be sent to Crunch -- the API still
+needs some metadata-driven rewrites (subvariable -> parent+axes, category
```
This is an interesting point, we have something similar happening in automation in which even after having an alias based expression, we still need to do a 2nd pass for the exact same motives.

This makes me think if we should start doing some of this in the API instead of demanding consumers (Scrunch, Crunch Automation) having to deal with this as well.

jjdelc commented on scrunch/expressions.py:
```diff
@@ -169,22 +170,16 @@ def r(lower, upper):
     return list(range(lower, upper + 1))
 
 
-def parse_expr(expr, platonic=False):
```
<3

jjdelc commented on scrunch/expressions.py:
```diff
@@ -505,80 +495,95 @@ def get_subvariables_resource(var_url, var_index):
+
+    Aliases are NOT converted to URLs -- the API accepts alias-based `var`
+    terms directly. Alias existence is validated here as a side-effect of
+    the metadata lookups.
```
Great docstring! Thanks for the clarity!

jjdelc reviewed (approved):
Thanks so much for this work! Great reworking, mustn't been easy to understand all of this to make this improvements. 🚀

slobodan-ilic committed:
Bump version to 0.18.22
